### PR TITLE
ECO-3547: Easycredit payment method address filter

### PR DESCRIPTION
**Pull request by denis-gnusov**
- Developer(s): @denis-gnusov

- Ticket: https://spryker.atlassian.net/browse/ECO-3547

#### Please confirm

- [ ] No new OS components - or they have been approved by legal department

#### Documentation

- [ ] Functional documentation provided or in progress?
- [ ] Integration guide for projects provided or not needed?
- [ ] Migration guides for all contained majors provided or not needed?

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   Heidelpay             | minor                 |                       |

-----------------------------------------

##### Change log

Improvements

- Adjusted `HeidelpayFacade::filterPaymentMethods()` to filter out EasyCredit method if billing address is not the same as shipping address.
- Impacted `HeidelpayPaymentMethodFilterPlugin` with facade changes.
- Introduced `HEIDELPAY:CONFIG_HEIDELPAY_EASYCREDIT_CRITERIA_COUNTRY_ISO_CODES` environment config to set allowed counties for EasyCredit payment method.

### CODE REVIEW

File: src/SprykerEco/Zed/Heidelpay/Business/Payment/PaymentMethodFilter.php [hunk 1]
```diff
@@ -17,6 +17,8 @@
 
 class PaymentMethodFilter implements PaymentMethodFilterInterface
 {
+    protected const ISO2_COUNTRY_DE = 'DE';
```
Comment (alex-galych) [inline src/SprykerEco/Zed/Heidelpay/Business/Payment/PaymentMethodFilter.php:20]: Move to the module configuration or to the environment-specific configuration

File: src/SprykerEco/Shared/Heidelpay/Transfer/heidelpay.transfer.xml [hunk 2]
```diff
@@ -385,6 +385,7 @@
         <property name="state" type="string"/>
         <property name="zipCode" type="string"/>
         <property name="salutation" type="string"/>
+        <property name="isDefaultBilling" type="bool"/>
```
[the remark this answers is not in the document]
Comment (denis-gnusov) [reply src/SprykerEco/Shared/Heidelpay/Transfer/heidelpay.transfer.xml:388]: I need it to have a transfer object constant.

File: src/SprykerEco/Shared/Heidelpay/Transfer/heidelpay.transfer.xml [hunk 3]
```diff
@@ -58,7 +58,7 @@
     </transfer>
 
     <transfer name="HeidelpayCreditCardPaymentOptions">
-        <property name="optionsList" type="HeidelpayPaymentOption[]"/>
+        <property name="optionsList" type="HeidelpayPaymentOption[]" singular="optionsList"/>
```
Comment (aleksey-kotsuba) [inline src/SprykerEco/Shared/Heidelpay/Transfer/heidelpay.transfer.xml:61]: Singular is invalid as well as in transfers below.

File: src/SprykerEco/Zed/Heidelpay/Business/Payment/PaymentMethodFilter.php [hunk 4]
```diff
@@ -77,18 +77,40 @@ protected function isPaymentMethodHeidelpayEasyCredit(PaymentMethodTransfer $pay
     }
 
     /**
-     * @param float $grandTotal
+     * @param \Generated\Shared\Transfer\QuoteTransfer $quoteTransfer
+     *
+     * @return bool
+     */
+    protected function isQuoteValidForEasyCredit(QuoteTransfer $quoteTransfer): bool
```
Comment (aleksey-kotsuba) [inline src/SprykerEco/Zed/Heidelpay/Business/Payment/PaymentMethodFilter.php:84]:
```suggestion
    protected function isEasyCreditAllowed(QuoteTransfer $quoteTransfer): bool
```

File: src/SprykerEco/Zed/Heidelpay/HeidelpayConfig.php [hunk 5]
```diff
@@ -178,6 +178,14 @@ public function getEasycreditCriteriaGrandTotalLessThan(): float
         return $this->get(HeidelpayConstants::CONFIG_HEIDELPAY_EASYCREDIT_CRITERIA_GRAND_TOTAL_LESS_THAN);
     }
 
+    /**
+     * @return string[]
+     */
+    public function getEasycreditCriteriaCountryIsoCodes(): array
```
Comment (aleksey-kotsuba) [inline src/SprykerEco/Zed/Heidelpay/HeidelpayConfig.php:184]:
```suggestion
    public function getEasyCreditAllowedCountries(): array
```

File: src/SprykerEco/Zed/Heidelpay/Business/Payment/PaymentMethodFilter.php (hunk 4 above)
Comment (alex-galych) [reply src/SprykerEco/Zed/Heidelpay/Business/Payment/PaymentMethodFilter.php]: Fixed.

File: src/SprykerEco/Zed/Heidelpay/HeidelpayConfig.php (hunk 5 above)
Comment (alex-galych) [reply src/SprykerEco/Zed/Heidelpay/HeidelpayConfig.php]: Fixed.

File: src/SprykerEco/Shared/Heidelpay/Transfer/heidelpay.transfer.xml (hunk 3 above)
Comment (alex-galych) [reply src/SprykerEco/Shared/Heidelpay/Transfer/heidelpay.transfer.xml]: Fixed.

**Pull request merged**